openedx/openedx-platform
# feat: due date reminder

## Description

This PR adds a command to send reminder emails for self-paced courses with relative due dates enabled. The command takes in a argument `due-in` which allows us to send reminders for subsections due within specified number of days.

## Testing instructions

* Setup master devstack
* Enable [SelfPacedRelativeDateConfig](http://localhost:18000/admin/course_date_signals/selfpacedrelativedatesconfig/) globally.
* Add waffle flag `studio.custom_relative_dates` to allow adding custom relative dates to graded subsections.
* Start LMS and CMS using `make {lms,cms}-up`.
* Run `make lms-migrate` to add enqueue and deliver columns for new schedule.
* Open [Schedules config](http://localhost:18000/admin/schedules/scheduleconfig/) and add an entry for `localhost:18000` site and enable all flags.
* Go to demo course in studio and update start date to some date in past (around 2 months) and change pacing to `self-paced` (for this you might need to update start_date to some date in future and revert back).
* Login to LMS using `edx` user and enroll in the demo course.
* Go to [schedules](http://localhost:18000/admin/schedules/schedule/) and update edx user's start_date to some recent date within the past one week.
* Open LMS shell using `make lms-shell` and run below command, play with date and due-in parameter and see the output. It should print email with subsections having due date falling within the days passed via `due-in` parameter.
  ```bash
  python manage.py lms send_course_due_date_reminders localhost:18000 --due-in 7
  ```
* Also, it should only send email if `SelfPacedRelativeDateConfig` is enabled and if course pacing is set to `self-paced`.
* Re-run the command by updating `Due in` field for a graded subsection:

![image](https://github.com/openedx/edx-platform/assets/10894099/5b692e70-ff52-41c0-a3fd-428e80feafcd)

**Example output:**

```log
------- EMAIL -------
To: edx@example.com
From: Demonstration Course
Subject: Important course due dates coming up
Body:
Hi there! We wanted to remind you of some important due dates coming up for your course, Demonstration Course, that you need to be aware of. The due dates are coming up in the following subsections:


  * Lesson 1 - Getting Started (28 Jul 2023)

  * Homework - Question Styles (07 Aug 2023)

  * Lesson 2 - Let's Get Interactive! (02 Aug 2023)

  * Homework - Labs and Demos (02 Aug 2023)

  * Homework - Essays (02 Aug 2023)

  * Lesson 3 - Be Social (08 Aug 2023)

  * Homework - Find Your Study Buddy (08 Aug 2023)

  * More Ways to Connect (08 Aug 2023)


Happy learning!
-------  END  -------
```

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

**Doubts:**

* Why is [access_duration](https://github.com/open-craft/edx-platform/blob/a0de151f7ba8b5d4cad7101b2c08c384235010a8/openedx/core/djangoapps/course_date_signals/utils.py#L31-L31) limited to max duration of 18 weeks?
* Copied relative week due date logic from [course_date_signals handler](https://github.com/open-craft/edx-platform/blob/43765ef72c16db88f923fbc74d3897cf54744b0e/openedx/core/djangoapps/course_date_signals/handlers.py#L146-L146). It sets minimum of 1 week for each section. Not sure if it is valid.
* Some tests in [schedules app](https://github.com/open-craft/edx-platform/blob/5300caa208f9a64ae0c7c3062c76b6254d4de706/openedx/core/djangoapps/schedules/management/commands/tests/test_send_course_update.py#L31-L31) are not running due to invalid condition in skip check. Many tests fail as soon as we fix the condition, we probably cannot fix it now but creating a issue might help.
  ```python
  # Currently it is set to.
  'openedx.core.djangoapps.schedules.apps.SchedulesConfig' in settings.INSTALLED_APPS
  # Should be below as it is set this way in common settings.
  'openedx.core.djangoapps.schedules' in settings.INSTALLED_APPS
  ```

Private-ref: [BB-7407](https://tasks.opencraft.com/browse/BB-7407)
